### ISSUE TITLE
Drill to detail on month only returns records from the first day of month

### ISSUE THREAD

**Issue by sfirke** [apache/superset#23847]
Drill to detail on a month should return all records from that month.  It returns only records from the first day.

#### How to reproduce the bug

1. Create an aggregate table chart with a datetime variable as a dimension
2. Select Time Grain = Month
3. On dashboard, Drill to Detail by Month.

### Expected results

See all results for that month (or month & other variables, if drilling by all).

### Actual results

Only records matching the first day of the month are returned.  I ran a trace on my data warehouse and can see the SQL query is only asking for the first day of the month:
```
SELECT TOP 50000 COUNT(*) AS [COUNT(*)]
FROM dbo.[myDataTable]
WHERE date_completed >= CONVERT(DATE, '2023-01-01', 23)
  AND date_completed < CONVERT(DATE, '2023-04-27', 23)
  AND date_completed = N'2023-03-01T00:00:00.000Z'
```

(The first two `date_completed` lines come from a dashboard native filter time range).

#### Screenshots

![image](https://user-images.githubusercontent.com/7569808/234896606-46a97389-11c5-4c92-af26-3f48f4a1bbc8.png)

### Environment

(please complete the following information):

- browser type and version: Firefox
- superset version: 2.1.0

**Comment by rusackas** [comment]
Well, you know the "we don't support 2.x" party line I'm using ;) Is this still a thing in 3.x?

**Comment by sfirke** [comment]
@rusackas Yep I just updated the original post to indicate I've validated it in 3.1.1.

**Comment by rusackas** [comment]
I assume this one is still happening in 4.x, but it's getting a little stale. Seems not many people are running into this.

**Comment by sfirke** [comment]
Yes, I confirmed that this is still happening in 4.0.2.  I updated the issue title to be more specific that this only affects table charts.

I tried to figure out where in the codebase changes would need to be made, I wonder if it's here since it only affects table charts and mine would have a pivot operation: https://github.com/apache/superset/blob/5bebcf61fefe2af7ed32f822ad1d86589ff4c9d2/superset/common/query_context_factory.py#L157

Also kind of wonder if it could be SQL Server-specific, maybe someone using a different data warehouse could try to validate.

**Comment by nemosdo** [comment]
I have the same problem, please let me know if you have any progress.

**Comment by sfirke** [comment]
I just validated that it persists in 4.1.2, I'll update the top post.

@nemosdo what is your data warehouse? I'd like to know if this is specific to SQL Server or affects all kinds of DBs.

**Comment by Samuelinto** [comment]
Good afternoon,
My group and I are students in the Software Engineering (CSCD01) course at the University of Toronto. We were wondering if this issue is still open, and if we could be assigned to fix it. We were able to replicate the issue, and have looked into ways we can fix it. 

Thanks.
@Laksh03 @nafizul02 @aidand9

**Comment by sfirke** [comment]
@Samuelinto that would be great! I would value that contribution and I think it seems like a good first issue. Please let me know if you have any questions as you go.

**Comment by Samuelinto** [comment]
Hi @sfirke thanks for your response. We were able to look deeper into the code base and found out that the issue persists within different types of databases and different time grains. As you alluded above, what appears to be happening is that the SQL query uses the truncated values provided by the time grain to filter the entries, which is causing only the rows that have the time set to exactly the first day of the month at midnight to be shown. The solution we came up with was adding a different conditional filter for datetime columns that uses the granularity to set an upper and lower bound to the queries. We're currently working on implementing the fix, but we wanted to get your opinion on if this seems like a correct approach. We also noticed that different charts build the drill-to-detail filter payloads in different ways that don't seem to suffer from this issue. Do you know if there is a reason for this difference? Thank you.

TableChart.tsx payload builder:

<img width="623" height="269" alt="Image" src="https://github.com/user-attachments/assets/3aa67bd7-60cf-4745-aec7-cc71405e1102" />

EchartsMixedTimeseries.tsx payload builder:

<img width="429" height="286" alt="Image" src="https://github.com/user-attachments/assets/01892cbd-3123-493d-beb9-0e56b4f8c07b" />

**Comment by sfirke** [comment]
@Samuelinto good stuff. I think your suggested approach of using granularity values to build upper and lower bounds seems good. 

That's intriguing, what you shared about the different ways the drill-to-detail payloads are constructed! I would say use your judgment to discern whether it has to be that way for some reason, or if not, then is this something we can DRY out by abstracting it into its own function we can repeat? And maybe solve the problem that way?

**Comment by Samuelinto** [comment]
Hi @sfirke, we just wanted to let you know that my team and I recently created a [pull request](https://github.com/apache/superset/pull/36318) with the changes discussed above. Please let us know if you have any feedback about it, and we hope our fix is able to resolve the issue.

**Comment by jaymasiwal** [comment]
Hi @sfirke — it looks like the PR from the university team has been inactive for a few months. If no one else is currently working on it, I’d love to pick this up!

My plan is to start by abstracting the payload builder from the Echarts implementation to handle the time grain correctly, just like you suggested.

Let me know if you are good with me taking ownership of this, and I'll start putting a PR together.

**Comment by sfirke** [comment]
@jaymasiwal sure, it seems like a good one to work on! @Samuelinto did a good job identifying the problem but then things have gone quiet.

Can you start by spinning up the latest Superset and creating a case of this bug to validate it? Then you will have the "before" for your before/after screenshots and you will know for sure you have fixed the bug because the behavior changes.

**Comment by sfirke** [comment]
Err sorry I commented before seeing that @Samuelinto and their team had submitted a PR to fix this, https://github.com/apache/superset/pull/36318. In that case I'd say hold on and let's see if @Samuelinto can answer if they want to keep working on that PR.

If they say no, or don't answer, you could take it over and respond/address reviewer feedback there as the next step. Let's give it a week for the original author to answer.

Given that this is mostly solved it would set us back to start a new PR and lose the existing work and thought that's gone into solving this.

**Comment by jaymasiwal** [comment]
Sounds good, @sfirke ! I'll check back next week. If they aren't able to return to it, I'd be glad to pick up where they left off so their work doesn't go to waste. I'll grab a different issue for now.

**Comment by jaymasiwal** [comment]
@sfirke, It’s been a week with no response, so I’m stepping in to get this across the finish line as discussed. I'll pull down their branch, address the remaining Copilot/bot feedback (specifically the time range boundary logic and null handling), add the missing tests, and open a fresh PR with a green CI.

**Comment by Samuelinto** [comment]
Good afternoon,
Sorry for the late reply. After we opened the PR,  @msyavuz had indicated that this error had been resolved in the past, though a regression might have occurred. Afterwards, @rusackas seemed to have taken control of the PR and made some changes and tests for it, so we thought that the issue had been resolved. We are willing to provide our support to resolve this issue, though we are unsure as to what its missing. Let us know if you need anything.

**Comment by jaymasiwal** [comment]
@Samuelinto No worries! I did a deep dive into the current master branch and confirmed that the regression is definitely still there (temporal drill-downs are still emitting exact matches rather than ranges). I've opened a fresh PR (#39668) that resolves the boundaries and handles the null temporal edge cases. All tests are passing!

**Comment by sfirke** [comment]
Fixed by #39668 - just over three years old.

**Comment by Samuelinto** [comment]
Glad to see that this issue is finally resolved. We hope that our contribution was useful in resolving the issue, and thank you @jaymasiwal for taking over and fixing the edge cases.

**Comment by sfirke** [comment]
Yes thanks @Samuelinto for getting this work rolling!

**Comment by jaymasiwal** [comment]
Thank you @Samuelinto and @sfirke ! It was great to build on the initial work and ensure those temporal edge cases were finally handled. Happy to see this resolved after so long! Looking forward to contributing more to the project.
